pasta/base/scope.py

```diff
@@ -95,7 +95,7 @@ def visit_FunctionDef(self, node):
     self.visit_in_order(node, 'decorator_list')
     self.scope.define_name(node.name, node)
     try:
-      self.scope = Scope(self.scope)
+      self.scope = self.root_scope.create_scope(self.scope, node)
```

Comment:
Maybe I'm not following how your internal scope model works but, shouldnt you have to nest these with the parent rather than going all the way up to the root?

If you have:
```
def f(x):
   class C():
      def g(y):
          pass
```

Then g is in the scope of f.

Reply:
(Also so is C, so its not strictly the syntactic parent)

Reply:
The idea here is to find the scope created by a node. Given the root scope (which is all you get from analyzing a tree), I want to be able to find subscopes within that. There is a scope created by a `FunctionDef`, so I can look up that scope by querying for the node that created it. For nodes which don't create a scope, I return the scope they are in.

The scopes we have here are described as:
```
<root>: names = [f]
     f: names = [x, C]
     C: names = [g]
     g: names = [y]
```

but each subscope is still aware of its parents. E.g, `scope_g.lookup_name('f')` returns the correct `Name` reference.

Does that help clarify? I realize this could use more documentation and probably more thought...

Reply:
Ohh I guess I was misunderstanding what the root_scope referred to here